modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -589,10 +588,29 @@ private CompletableFuture<Void> performRebalanceOnRecovery(long revision) {
 
         try (Cursor<Entry> cursor = metaStorageMgr.prefixLocally(prefix, revision)) {
             CompletableFuture<?>[] futures = cursor.stream()
-                    .map(this::handleChangePendingAssignmentEvent)
+                    .map(pendingAssignmentEntry -> {
+                        if (LOG.isInfoEnabled()) {
+                            LOG.info(
+                                    "Missed pending assignments for key '{}' with revision {} discovered, performing recovery",
+                                    new String(pendingAssignmentEntry.key(), UTF_8),
+                                    revision
+                            );
+                        }
+
+                        // We use the Meta Storage recovery revision here instead of the entry revision, because
+                        // 'handleChangePendingAssignmentEvent' accesses some Versioned Values that only store values starting with
+                        // tokens equal to Meta Storage recovery revision. In other words, if the entry has a lower revision than the
+                        // recovery revision, there will never be a Versioned Value corresponding to its revision.
+                        return handleChangePendingAssignmentEvent(pendingAssignmentEntry, revision);
```

Comment:
Perhaps I'm missing something, but won't the `pendingAssignmentEntry` have the same revision as the `metaStorageMgr.recoveryFinishedFuture()`?

Reply:
no, its revision will _at most_ be equal to the recovery revision, but it doesn't have to be equal (remember that revision gets changed for any key update)

Reply:
I don't understand your comment a bit.
In your case, there is no need for an asynchronous start of recovery; you can read the keys from the metastore and then start processing them asynchronously, right?

Reply:
I need to read keys from the Meta Storage that were missed by Watch processing (e.g. while a node has been offline) and I can process them asynchonously